schema.c

```diff
@@ -4065,3 +4381,517 @@ sch_json_to_gnode (sch_instance * instance, sch_node * schema, json_t * json, in
+    {
+        char *str = sch_translate_get_wildcard_data_from_bottom (node, xlat_data->int_tree);
+        if (str)
+        {
+            GNode *new_nnode;
+            if (sch_translate_child_exists (nnode, str, &new_nnode))
+            {
+                nnode = new_nnode;
+                g_free (str);
+            }
+            else
+                nnode = g_node_append_data (nnode, str);
+        }
+
+        node1 = node1->children;
+    }
+
+    /* We should now be synchronized at a wildcard in both int and external trees,
+       add in fields after the wild card */
+    while (node1 && node1->children)
+    {
+        GNode *new_nnode;
+
+        if (nnode && sch_translate_child_exists (nnode, APTERYX_NAME (node1), &new_nnode))
+            nnode = new_nnode;
+        else
+            nnode = g_node_append_data (nnode, g_strdup (APTERYX_NAME (node1)));
+
+        node1 = node1->children;
+    }
+
+    /* Now copy the exact match node to the output tree */
+    copy = g_node_copy_deep (node, copy_node_data, NULL);
+    sch_translate_mark_node (node_table, node);
+    sch_translate_mark_children (node_table, node);
+
+    /* Check if a field name change is required */
+    if (node1)
+    {
+        if (g_strcmp0 (APTERYX_NAME (copy), APTERYX_NAME (node1)))
+        {
+            g_free (copy->data);
+            copy->data = g_strdup (APTERYX_NAME (node1));
+        }
+    }
+
+    nnode = g_node_append (nnode, copy);
+}
+
+static void
+sch_translate_output_matches (sch_instance * instance, GNode **ret_node, GNode *node,
+                              GNode *int_node, GHashTable *node_table, int flags,
+                              sch_xlat_data *xlat_data)
+{
+    GNode *child;
+    GNode *next_child;
+
+    if (g_strcmp0 ((char *) node->data, (char *) int_node->data) &&
+        g_strcmp0 ((char *) int_node->data, "*"))
+        return;
+
+    if (!int_node->children && g_hash_table_lookup (node_table, node))
+        return;
+
+    int_node = int_node->children;
+    if (!int_node)
+    {
+        if (xlat_data->lua_module && node->children)
+            sch_translate_data (instance, node, xlat_data, flags, false);
+
+        /* We have an exact match, translate the path into the output tree */
+        sch_translate_merge_output (ret_node, node, node_table, xlat_data);
+
+        return;
+    }
+
+    for (child = node->children; child; child = next_child)
+    {
+        next_child = child->next;
+        sch_translate_output_matches (instance, ret_node, child, int_node, node_table,
+                                      flags, xlat_data);
+    }
+}
+
+GNode *
+sch_translate_output (sch_instance * instance, GNode *node, int flags, void *xlat_data)
+{
+    sch_xlat_data *x_data = (sch_xlat_data *) xlat_data;
+    sch_xlat_root *root;
+    GList *list;
+    GHashTable *node_table = NULL;
+    int num_nodes;
+    GNode *ret_node = NULL;
+
+    if (!instance->xlat_hash_table || !node)
+        return node;
+
+    if (!x_data)
```

Comment:
does this avoid any more processing if there was no input translation???

Reply:
There may be a many to one mapping from external models to an internal model. Without the x_data to determine which external model we are dealing with, output translation will not work reliably.